Introduce RemoteFrame support for testRunner.dumpChildFramesAsText

achristensen07 opened the pull request:
#### 46520106b8b99191f7af6b78883c380f883178f7
<pre>
Introduce RemoteFrame support for testRunner.dumpChildFramesAsText
<a href="https://bugs.webkit.org/show_bug.cgi?id=271724">https://bugs.webkit.org/show_bug.cgi?id=271724</a>
<a href="https://rdar.apple.com/125437971">rdar://125437971</a>

Reviewed by Charlie Wolfe.

This is the first of at least two steps to get tests like http/tests/security/window-properties-pass.html
passing with --site-isolation.  If the frame text is not in the current process, send a message to the
process that contains the frame to get its text.  FrameTree::uniqueName is not hooked up correctly so
the test output isn&apos;t identical yet, but this makes more than 0 output.

The separate asDocumentElement test wasn&apos;t actually useful because it was redundant with the same
test done in WebFrame::innerText, so it was removed.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::renderTreeAsTextForTesting):
(WebKit::WebPageProxy::frameTextForTesting):
(WebKit::WebPageProxy::renderTreeAsText): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKBundleFrameCopyInnerText): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageCopyFrameTextForTesting):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::renderTreeAsText):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::innerText const):
(WebKit::WebFrame::frameTextForTesting):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameTextForTestingIncludingSubframes):
(WebKit::WebPage::renderTreeAsTextForTesting):
(WebKit::WebPage::frameTextForTesting):
(WebKit::WebPage::renderTreeAsText): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp:
(WTR::InjectedBundlePage::dumpAllFramesText):
(WTR::InjectedBundlePage::dump):
(WTR::hasDocumentElement): Deleted.
(WTR::dumpFrameText): Deleted.
(WTR::dumpDescendantFramesText): Deleted.

Canonical link: <a href="https://commits.webkit.org/276723@main">https://commits.webkit.org/276723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8fcdac3dea5081072b4be079b65455da22a3583

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45472 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48006 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48138 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41480 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47779 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28838 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21991 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21679 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39234 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18398 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19081 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/40330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3518 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41778 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40647 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49867 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20457 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16986 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44342 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21765 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43171 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22124 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6327 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21452 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1318,4 +1318,36 @@ bool WebFrame::isFocused() const
+
+    // To keep things tidy, strip all trailing spaces: they are not a meaningful part of dumpAsText test output.
+    // Breaking the string up into lines lets us efficiently strip and has a side effect of adding a newline after the last line.
+    if (String text = innerText(); !!text) {
```
Returning early if this is not true would match previous behavior.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1318,4 +1318,36 @@ bool WebFrame::isFocused() const
+    for (auto* child = m_coreFrame->tree().firstChild(); child; child = child->tree().nextSibling()) {
+        if (is<RemoteFrame>(*child)) {
+            builder.append(m_page->sendSync(Messages::WebPageProxy::FrameTextForTesting(child->frameID())).takeReplyOr("Sending WebPageProxy::FrameTextForTesting failed"_s));
+        } else if (RefPtr childWebFrame = fromCoreFrame(*child)) {
```
Similar here, checking `!childWebFrame->innerText().isNull()` in this if statement would match previous behavior.

achristensen07 committed:
Introduce RemoteFrame support for testRunner.dumpChildFramesAsText
https://bugs.webkit.org/show_bug.cgi?id=271724
rdar://125437971

Reviewed by Charlie Wolfe.

This is the first of at least two steps to get tests like http/tests/security/window-properties-pass.html
passing with --site-isolation.  If the frame text is not in the current process, send a message to the
process that contains the frame to get its text.  FrameTree::uniqueName is not hooked up correctly so
the test output isn't identical yet, but this makes more than 0 output.

The separate asDocumentElement test wasn't actually useful because it was redundant with the same
test done in WebFrame::innerText, so it was removed.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::renderTreeAsTextForTesting):
(WebKit::WebPageProxy::frameTextForTesting):
(WebKit::WebPageProxy::renderTreeAsText): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKBundleFrameCopyInnerText): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageCopyFrameTextForTesting):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::renderTreeAsText):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::innerText const):
(WebKit::WebFrame::frameTextForTesting):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameTextForTestingIncludingSubframes):
(WebKit::WebPage::renderTreeAsTextForTesting):
(WebKit::WebPage::frameTextForTesting):
(WebKit::WebPage::renderTreeAsText): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp:
(WTR::InjectedBundlePage::dumpAllFramesText):
(WTR::InjectedBundlePage::dump):
(WTR::hasDocumentElement): Deleted.
(WTR::dumpFrameText): Deleted.
(WTR::dumpDescendantFramesText): Deleted.

Canonical link: https://commits.webkit.org/276723@main
